### DISCUSSION TITLE
Is it possible to pipe buffer contents to external program?

### DISCUSSION THREAD

**Discussion by ianliu** [zed-industries/zed#12007] [Feature Requests]
In vim you can select a range of text and pipe its content to external programs, choosing what to do with its output:

1. display the output in vim logs (messages)
2. replace the selection with the output

This model is very powerful and I was wondering if there's anything like it in Zed.

**Comment by olejorgenb** [comment]
Using a clipboard CLI get's you 70% the way. `<ctrl-c>`  open-terminal `clip | tr : "\n" | clip` close-terminal `<ctrl-v>`

**Comment by JonGretar** [comment]
This all almost exactly the same in zed. With the `!` command. A few examples:

 1. Type `:` to get in the command prompt and type `!wc %` (% inserts current file name) and you open a terminal counting the words in the current files. 
 1. Type `:` to get in the command prompt and type `.!whoami` (notice the dot) and you replace current line with your username . 
 2. Type `:` to get in the command prompt and type `r!whoami`  and you insert a line below current with your username . 
 4. Type `:` to get in the command prompt and type `%! grep TODO` and your buffers contents will be replaced with the lines containing "TODO"
 5. Select a bunch of lines and type `:` to get in the command prompt. You will see `'<,'>` and add to that `!wc` (So it says `'<,'>!wc`) and your selected text will be replaced with the word count of the selected text.
 6. Select a bunch of lines and type `:` to get in the command prompt. You will see `'<,'>` and add to that `r!wc` and your word count of the selected text will be appended to the cursor location.

There is a bunch more options. Relative lines prefixes and so on.

**Reply by olejorgenb** [reply]
Not able to get this to work. The vim bindings bind `:` to `command_palette::Toggle`, and opening this and typing "!wc %" does nothing.

So maybe this only work in vim mode ? :/

**Reply by JonGretar** [reply]
Ahh. Possibly. Given your question I assumed that you already were in vim mode.
Should be relatively easy to ask for this to be added in other modes.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
